### PR TITLE
[WS1][kernels] Batch-invariant logprob (CUDA)

### PR DESCRIPTION
**Pull request by KJLdefeated**
# [WS1][kernels] SM90 Batch-Invariant logprob CUDA Kernel

## Summary

Adds a Hopper (SM90) TMA CUDA kernel for the batch-invariant selected-token log-probability operator introduced in #199  plus a Native/Triton/CUDA benchmark. The operator computes, from **already materialized** logits:
```text
logp[t] = logits[t, target_ids[t]] - logsumexp(logits[t, :])
```

The kernel streams the vocab through shared memory with TMA bulk-tensor copies and folds it into a running online log-sum-exp, so it never materializes the `[N, V]` softmax and its extra forward memory is ~0. It is registered as the
top-priority CUDA backend on Hopper and falls back cleanly everywhere else.

| Path | Status |
| --- | --- |
| CUDA SM90 forward (`csrc/cuda/batch_invariant_logp_kernel_sm90.cu`) | TMA online-softmax, bf16 + fp32, one CTA per row. New. |
| Python op (`BatchInvariantLogpSM90Op`) | Autograd wrapper; TMA forward + tile-wise backward. New. |
| Backward | Reuses #199's Triton tile-wise backward via the forward-saved `lse` (no `[N, V]` materialization). |
| Registry | Hardware-gated: inserted at the front of the CUDA `batch_invariant_logp` list only when the SM90 symbol is compiled on `cc_major == 9`. |
| Fallback | Non-bf16/fp32, unaligned vocab row stride, or non-Hopper → Triton, else native. |
| Benchmark | `benchmarks/benchmark_batch_invariant_logp.py` compares Native/Triton/CUDA. New. |

This builds on the open branch for #199 (Native + Triton). Reviewed after #199 merges.

## Implementation

- **New kernel** `csrc/cuda/batch_invariant_logp_kernel_sm90.cu`.
- **Binding** in `csrc/ops.cpp` and **build source** in `setup.py`'s `sm90_srcs`; stub in `rl_engine/_C.pyi`.
- **Python op** `rl_engine/kernels/ops/cuda/loss/batch_invariant_logp.py`:
  - `_BatchInvariantLogpSM90Function` forward calls the compiled kernel;
  - backward reuses Triton `_batch_invariant_logp_bwd_kernel`.
- **Registry** `rl_engine/kernels/registry.py`.
- **Docs/benchmark**: updated `docs/operators/batch-invariant-logp.md`; added the benchmark script.

## Validation environment

| Item | Value |
| --- | --- |
| GPU | NVIDIA H200 (Hopper, SM90, compute capability 9.0) |
| Driver | 550.127.08 (CUDA driver 12.4) |
| CUDA toolkit / `nvcc` | 12.8, V12.8.93 (conda `nvidia` channel) |
| PyTorch | 2.11.0+cu128 |
| Python | 3.11.15 |
| Host compiler | gcc 12.5.0 |
| Extension | `KERNEL_ALIGN_FORCE_SM90=1`, `_C.batch_invariant_logp_sm90` available |
| Build vars | `TORCH_CUDA_ARCH_LIST=9.0a`, `KERNEL_ALIGN_DEV_RPATH=1` |

### Unit coverage

```bash
python -m pytest tests/test_batch_invariant_logp.py -q -rs
```

Result: **67 passed** (14 new SM90-kernel tests: correctness fp32/bf16, large and unaligned vocab, single-token, 3-D, bitwise batch-invariance, backward, ignore-index, fp16 fallback; plus the registry-dispatch test now accepting the CUDA op).

```bash
python -m pytest tests/test_batch_invariant_logp.py -k CudaSM90 -q
```

Result: **14 passed**. Registry selects `BatchInvariantLogpSM90Op` on the H200:

```text
Detected TMA-capable architecture (SM90); ...
Successfully linked to precompiled _C.batch_invariant_logp_sm90 kernel.
```

### Vocab sweep (bf16, vs `log_softmax + gather`)

Correct on both TMA-aligned and unaligned (tail-path) vocab sizes:

| V | 256 | 300 | 400 | 1024 | 50257 | 128256 | 151936 |
| --- | --- | --- | --- | --- | --- | --- | --- |
| max abs err | 0 | 0 | 4.8e-7 | 4.8e-7 | 9.5e-7 | 9.5e-7 | 9.5e-7 |

## gtest (based on #197 )
Common prefix: `python scripts/check_operator.py --op batch_invariant_logp --candidate cuda-sm90 --device cuda --arch-key sm90`
### All Pass
Used config:
```
# initial forward + backward
... --dtype bf16 --batch 4 --seq 8 --vocab 1024
... --dtype bf16 --batch 4 --seq 8 --vocab 1024 --check-grad

# dtype/shape sweep
... --dtype fp32 --batch 2 --seq 16 --vocab 2048
... --dtype bf16 --batch 8 --seq 32 --vocab 4096  --check-grad
... --dtype bf16 --batch 1 --seq 2  --vocab 152064 --check-grad
... --dtype fp32 --batch 2 --seq 4  --vocab 512 --input-mode constant --constant-value 0.5 --token-value 3

# batch/seq shape sweep (forward)
... --dtype bf16 --vocab 1024 --batch 1   --seq 1
... --dtype bf16 --vocab 1024 --batch 1   --seq 128
... --dtype bf16 --vocab 1024 --batch 3   --seq 7
... --dtype bf16 --vocab 1024 --batch 16  --seq 1
... --dtype bf16 --vocab 1024 --batch 32  --seq 64
... --dtype bf16 --vocab 1024 --batch 64  --seq 128
... --dtype bf16 --vocab 1024 --batch 128 --seq 4

# batch/seq shape sweep (forward + backward)
... --dtype bf16 --vocab 2048 --batch 2  --seq 3  --check-grad
... --dtype bf16 --vocab 2048 --batch 5  --seq 13 --check-grad
... --dtype bf16 --vocab 2048 --batch 17 --seq 31 --check-grad
... --dtype fp32 --vocab 4096 --batch 9  --seq 11 --check-grad

# large row counts: 1024 / 4096 / 8192 rows (batch*seq), --check-grad
... --dtype bf16 --vocab 2048 --batch 1024 --seq 1   --check-grad
... --dtype bf16 --vocab 2048 --batch 8    --seq 128 --check-grad
... --dtype bf16 --vocab 2048 --batch 32   --seq 32  --check-grad
... --dtype bf16 --vocab 2048 --batch 4096 --seq 1   --check-grad
... --dtype bf16 --vocab 2048 --batch 32   --seq 128 --check-grad
... --dtype bf16 --vocab 2048 --batch 64   --seq 64  --check-grad
... --dtype bf16 --vocab 2048 --batch 8192 --seq 1   --check-grad
... --dtype bf16 --vocab 2048 --batch 64   --seq 128 --check-grad
... --dtype bf16 --vocab 2048 --batch 128  --seq 64  --check-grad
... --dtype fp32 --vocab 4096   --batch 128 --seq 64  --check-grad
... --dtype bf16 --vocab 152064 --batch 64  --seq 128 --check-grad
```

## Benchmarks

`python benchmarks/benchmark_batch_invariant_logp.py` — bf16, H200, 20 iters + 5 warmup. Memory is peak extra above baseline (MB).

### Forward

| shape (N x V) | native ms | triton ms | **cuda ms** | cuda vs native | cuda vs triton | native MB | triton MB | cuda MB |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 4096x32768 | 1.341 | 0.148 | **0.090** | 14.8x | 1.64x | 1536 | 0 | 0 |
| 4096x128256 | 4.964 | 0.566 | **0.323** | 15.4x | 1.75x | 6012 | 0 | 0 |
| 4096x151936 | 5.908 | 0.669 | **0.384** | 15.4x | 1.74x | 7122 | 0 | 0 |
| 8192x128256 | 9.904 | 1.056 | **0.597** | 16.6x | 1.77x | 12024 | 0 | 0 |

### Forward + backward

| shape (N x V) | native ms | triton ms | **cuda ms** | cuda vs native | cuda vs triton | native MB | triton MB | cuda MB |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 4096x32768 | 3.549 | 0.463 | **0.407** | 8.7x | 1.14x | 1792 | 512 | 512 |
| 4096x128256 | 13.173 | 1.750 | **1.510** | 8.7x | 1.16x | 7014 | 2004 | 2004 |
| 4096x151936 | 15.632 | 2.072 | **1.788** | 8.7x | 1.16x | 8310 | 2376 | 2376 |
| 8192x128256 | 26.211 | 3.416 | **2.955** | 8.9x | 1.16x | 14028 | 4008 | 4008 |

## Files

- `csrc/cuda/batch_invariant_logp_kernel_sm90.cu` (new)
- `rl_engine/kernels/ops/cuda/loss/batch_invariant_logp.py` (new)
- `benchmarks/benchmark_batch_invariant_logp.py` (new)
- `csrc/ops.cpp`, `setup.py`, `rl_engine/_C.pyi`, `rl_engine/kernels/registry.py`
- `tests/test_batch_invariant_logp.py`, `docs/operators/batch-invariant-logp.md`

## Summary by CodeRabbit

* **New Features**
  * Added a Hopper GPU-optimized path for batch-invariant selected-token log-probabilities.
  * Automatically selects the optimized backend when supported, with fallback support for other platforms and hardware.
  * Added forward and backward computation support, including ignored-target handling and input validation.
  * Added benchmarking for latency, memory usage, and backend speedups across configurable tensor sizes.

* **Documentation**
  * Added operator usage, backend behavior, accuracy expectations, benchmarks, and testing guidance.

* **Tests**
  * Added coverage for correctness, gradients, batch invariance, data types, fallback behavior, and Hopper-specific execution.

### CODE REVIEW

**Comment by Flink-ddd** [comment]
please resolve the code conflicts first, Thanks. PR: #199 merged.

**Comment by KJLdefeated** [comment]
@Flink-ddd merge conflict solved and CI are all pass. request for review, thx.

File: csrc/cuda/batch_invariant_logp_kernel_sm90.cu [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    const int smem_size = (SMEM_TILE * sizeof(nv_bfloat16)) + 16;
+    batch_invariant_logp_sm90_kernel<nv_bfloat16, NUM_WARPS><<<N, NUM_WARPS * 32, smem_size>>>(
+        logits_tmap, target.data_ptr<int>(),
+        reinterpret_cast<const nv_bfloat16 *>(logits.data_ptr<at::BFloat16>()),
+        logp.data_ptr<float>(), lse.data_ptr<float>(), N, V, ignore_index);
+}
+
+} // namespace
+
+std::vector<torch::Tensor> batch_invariant_logp_sm90_forward(torch::Tensor logits,
```
Comment (Flink-ddd) [inline csrc/cuda/batch_invariant_logp_kernel_sm90.cu:180]: Add a check to ensure logits.data_ptr() % 16 == 0 (in C++: reinterpret_cast<uintptr_t>(logits.data_ptr()) % 16 == 0). While you correctly verified the row stride alignment, Hopper TMA hardware also strictly requires the global base address to be 16-byte aligned; sliced tensors might be contiguous and have correct strides, but unaligned base pointers will trigger a CUDA illegal memory access fault.

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
Left a few comments, otherwise LGTM!

**Comment by maxiaosong1124** [comment]
@KJLdefeated Please address the issues mentioned in the comments above. If all are resolved, I am happy to approve.

**Comment by Flink-ddd** [comment]
please resolve the code conflicts and CI error, other LGTM!

**Comment by KJLdefeated** [comment]
@Flink-ddd @maxiaosong1124  Done.

**Review by maxiaosong1124 — APPROVED** [review]
LGTM!  thanks!

**Review by Flink-ddd — APPROVED** [review]
LGTM now, Thank you for update.
